core/shared/src/main/scala/fs2/internal/Algebra.scala

```diff
@@ -64,126 +85,140 @@ private[fs2] object Algebra {
-      case bound: FreeC.Bind[Algebra[F,O,?],x,Unit] =>
-        val fx = bound.fx.asInstanceOf[FreeC.Eval[Algebra[F,O,?],x]].fr
-        fx match {
-          case os: Algebra.Output[F,O] =>
-            pure(Some((os.values, bound.f(Right(())))))
-          case os: Algebra.Run[F,O,x] =>
-            try {
-              def asSegment(c: Catenable[Chunk[O]]): Segment[O,Unit] =
-                c.uncons.flatMap { case (h1,t1) => t1.uncons.map(_ => Segment.catenated(c.map(Segment.chunk))).orElse(Some(Segment.chunk(h1))) }.getOrElse(Segment.empty)
-              os.values.force.splitAt(chunkSize, Some(maxSteps)) match {
-                case Left((r,chunks,rem)) =>
-                  pure(Some(asSegment(chunks) -> bound.f(Right(r))))
-                case Right((chunks,tl)) =>
-                  pure(Some(asSegment(chunks) -> FreeC.Bind(segment(tl), bound.f)))
-              }
-            } catch { case NonFatal(e) => FreeC.suspend(uncons(bound.f(Left(e)), chunkSize)) }
-          case algebra => // Eval, Acquire, Release, OpenScope, CloseScope, GetScope
-            FreeC.Bind(
-              FreeC.Eval(algebra.asInstanceOf[Algebra[F,X,x]]), // O is phantom in these constructors so it is safe to case O to X
-              (e: Either[Throwable,x]) => uncons[F,X,O](bound.f(e), chunkSize)
-            )
-        }
-      case e => sys.error("FreeC.ViewL structure must be Pure(a), Fail(e), or Bind(Eval(fx),k), was: " + e)
-    }
-  }
+  def uncons[F[_],X,O](s: FreeC[Algebra[F,O,?],Unit], chunkSize: Int = 1024, maxSteps: Long = 10000): FreeC[Algebra[F,X,?],Option[(Segment[O,Unit], FreeC[Algebra[F,O,?],Unit])]] =
+    FreeC.Eval[Algebra[F,X,?],Option[(Segment[O,Unit], FreeC[Algebra[F,O,?],Unit])]](Algebra.Uncons[F,O,X](s, chunkSize, maxSteps))
 
 
   /** Left-folds the output of a stream. */
   def compile[F[_],O,B](stream: FreeC[Algebra[F,O,?],Unit], init: B)(f: (B, O) => B)(implicit F: Sync[F]): F[B] =
     F.delay(CompileScope.newRoot).flatMap { scope =>
-      compileScope[F,O,B](scope, stream, init)(f).attempt.flatMap {
-        case Left(t) => scope.close *> F.raiseError(t)
-        case Right(b) => scope.close as b
-      }
-    }
+    compileScope[F,O,B](scope, stream, init)(f).attempt.flatMap {
+      case Left(t) => scope.close *> F.raiseError(t)
+      case Right(b) => scope.close as b
+    }}
 
   private[fs2] def compileScope[F[_],O,B](scope: CompileScope[F], stream: FreeC[Algebra[F,O,?],Unit], init: B)(g: (B, O) => B)(implicit F: Sync[F]): F[B] =
-    compileLoop[F,O,B](scope, init, g, uncons(stream).viewL)
+    compileFoldLoop[F,O,B](scope, init, g, stream)
 
-  private def compileLoop[F[_],O,B](
+  private def compileUncons[F[_],X,O](
+    scope: CompileScope[F]
+    , s: FreeC[Algebra[F,O,?],Unit]
+    , chunkSize: Int
+    , maxSteps: Long
+  )(implicit F: Sync[F]): F[(CompileScope[F], Option[(Segment[O,Unit], FreeC[Algebra[F,O,?],Unit])])] = {
+    F.delay(s.viewL.get) flatMap {
```

Comment:
Do we need to wrap `s.viewL.get` in a delay constructor?

Reply:
Yes, there are situation that will SoE without it. I think one of them was `transform` stack safety